src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs

```diff
@@ -0,0 +1,104 @@
+
+namespace NuGet.VisualStudio.Implementation.Extensibility
+{
+    internal class NuGetProjectService : INuGetProjectService
+    {
+        private readonly IVsSolutionManager _solutionManager;
+
+        public NuGetProjectService(IVsSolutionManager solutionManager)
+        {
+            _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
+        }
+
+        public async Task<InstalledPackagesResult> GetInstalledPackagesAsync(Guid projectId, CancellationToken cancellationToken)
+        {
+            // Just in case we're on the UI thread, switch to background thread. Very low cost (does not schedule new task) if already on background thread.
+            await TaskScheduler.Default;
+
+            NuGetProject project = await _solutionManager.GetNuGetProjectAsync(projectId.ToString());
+            if (project == null)
+            {
+                return NuGetContractsFactory.CreateInstalledPackagesResult(InstalledPackageResultStatus.ProjectNotReady, packages: null);
+            }
+
+            var status = InstalledPackageResultStatus.Successful;
+            List<NuGetInstalledPackage> installedPackages;
+
+            if (project is BuildIntegratedNuGetProject packageReferenceProject)
+            {
+                var cacheContext = new DependencyGraphCacheContext();
+                var (packageSpecs, messages) = await packageReferenceProject.GetPackageSpecsAndAdditionalMessagesAsync(cacheContext);
+                if (messages?.Any(m => m.Level == LogLevel.Error) == true)
+                {
+                    status = InstalledPackageResultStatus.ProjectInvalid;
+                }
+
+                var packageSpec = packageSpecs.Single(s => s.RestoreMetadata.ProjectStyle == ProjectModel.ProjectStyle.PackageReference || s.RestoreMetadata.ProjectStyle == ProjectModel.ProjectStyle.ProjectJson);
+                var packagesPath = VSRestoreSettingsUtilities.GetPackagesPath(NullSettings.Instance, packageSpec);
+                FallbackPackagePathResolver pathResolver = new FallbackPackagePathResolver(packagesPath, VSRestoreSettingsUtilities.GetFallbackFolders(NullSettings.Instance, packageSpec));
+
+                var packageReferences = await project.GetInstalledPackagesAsync(cancellationToken);
+
+                installedPackages = packageReferences.Select(p => ToNuGetInstalledPackage(p, pathResolver))
+                    .ToList();
+            }
+            else if (project is MSBuildNuGetProject packagesConfigProject)
+            {
+                FolderNuGetProject packagesFolder = packagesConfigProject.FolderNuGetProject;
+
+                var packageReferences = await project.GetInstalledPackagesAsync(cancellationToken);
+
+                installedPackages = packageReferences.Select(p => ToNuGetInstalledPackage(p, packagesFolder))
+                    .ToList();
+            }
+            else
+            {
+                // unknown/unsupported project type
```

Comment:
then set the `status=Unknown` ?

Reply:
Maybe, but on the other hand if the project doesn't support NuGet, then we know that the list of NuGet packages the project has is empty. So returning "we don't know what packages are installed' isn't exactly true.  Having said that, it's possible that I didn't have an exhaustive if/else list of supported project types. I could just use the generic `NuGetProject`'s `GetInstalledPackages` to get the list of packages, but then I'll have no way of knowing where the expected package path is. I don't know if we should just use/assume global packages folder, return null for the package path, or return an error.

This is a scenario I didn't think of in the API spec and unfortunately wasn't brought up in the review. I'll think about it while working on the other feedback, but anyone else's comments would be appreciated.

Reply:
I didn't bring it up in the spec review because I didn't know you would implement it with "status=success" and "projects={empty}" when there's no supported projects. At a minimum, I think status should be something other than `Successful`.

> but then I'll have no way of knowing where the expected package path is

How are you getting a package path from an empty List?

Reply:
> How are you getting a package path from an empty List?

You misunderstood what I was trying to say. What I meant is that we already have a valid instance of `NuGetProject`, it's just we were unable to typecast it to one of the two more specific types. But the base `NuGetProject` has a `GetInstalledPackages` method that could return one or more packages. In that case, how can we determine the install path, given it's not one of the two "known" project types?

Thinking a bit more clearly after sleeping overnight: At this point in the code, we've already successfully got the `NuGetProject` instance out of the project system cache. Therefore we know it's a project type that supports NuGet, otherwise it never would have been in the project cache.  Looking at which classes implement `NuGetProject`, there are only 3 that do no extend either `MSBuildNuGetProject` or `BuildIntegratedNuGetProject`:
* `PackagesConfigNuGetProject`: This represents a `packages.config` file that is restored on the command line. Can't happen in VS, and `NuGetProjectService` is a VS service.
* `FolderNuGetPackage`: This represents the solution packages folder. It can't be a project in the solution.
* `ProjectKNuGetPackage`: This was a new project type that the ASP.NET Core team created during previews of .NET Core (.NET Core was called Project K before being renamed). Later, during the Visual Studio 2017 timeframe, trying to open a `xproj` would prompt to convert the project to a `csproj`. I don't believe that Visual Studio 2019 ever supported loading `xproj`, and therefore it should be impossible for this project type to be loaded any longer.

So, I need to update the comment, because "unsupported project type" is not correct, that can't be the case at the line where the comment is written. And with today's codebase, it's also impossible to get into this `else` block. Anything we put in here is currently purely theoretical, or a precaution for new project types to be added in the future.  Now I'm trying to decide if your suggestion is the best option, as the documentation for `Unknown` says that it's probably a bug, or if I should just throw, because it is a bug in our implementation.